src/query/executor.rs

```diff
@@ -0,0 +1,162 @@
+    fn into_bytes(self: Box<Self>) -> Vec<u8> {
+        // todo Result
+        let mut r = Vec::with_capacity(2 * self.len() * F::SIZE_IN_BYTES as usize);
+        for share in self.into_iter() {
+            let mut slice = [0u8; 32]; // we don't support fields > 16 bytes
+            let written = share.serialize(&mut slice).unwrap();
+            r.extend_from_slice(&slice[..written]);
+        }
+
+        r
+    }
+}
+
+#[cfg(any(test, feature = "test-fixture"))]
+async fn test_multiply<F: Field, S: Stream<Item = Vec<u8>> + Send + Unpin>(
+    ctx: SemiHonestContext<'_, F>,
+    mut input: S,
+) -> Vec<Replicated<F>> {
+    use crate::protocol::basics::SecureMul;
+    use crate::protocol::RecordId;
+    use futures_util::StreamExt;
+
+    let mut results = Vec::new();
+    while let Some(v) = input.next().await {
+        // convert bytes to replicated shares
+        let shares = v.chunks(2 * F::SIZE_IN_BYTES as usize).map(|chunk| {
+            // TODO fix with replicated serialization
+            let left = F::deserialize(&chunk[..=F::SIZE_IN_BYTES as usize]).unwrap();
+            let right = F::deserialize(&chunk[F::SIZE_IN_BYTES as usize..]).unwrap();
+
+            Replicated::new(left, right)
+        });
+
+        // multiply pairs
+        let mut a = None;
+        let mut record_id = 0_u32;
+        for share in shares {
+            match a {
+                None => a = Some(share),
+                Some(a_v) => {
+                    let result = ctx
+                        .clone()
+                        .multiply(RecordId::from(record_id), &a_v, &share)
+                        .await
+                        .unwrap();
+                    results.push(result);
+                    record_id += 1;
+                    a = None;
+                }
+            }
+        }
+
+        assert!(a.is_none());
+    }
+
+    results
+}
+
+#[allow(clippy::unused_async)]
+async fn ipa<F: Field, S: Stream<Item = Vec<u8>> + Send + Unpin>(
+    _ctx: SemiHonestContext<'_, F>,
+    _input: S,
+) -> Vec<Replicated<F>> {
+    todo!()
+}
+
+pub fn start_query<S: Stream<Item = Vec<u8>> + Send + Unpin + 'static>(
+    config: QueryConfig,
+    gateway: Gateway,
+    input: S,
+) -> JoinHandle<Box<dyn Result>> {
+    tokio::spawn(async move {
+        // TODO: make it a generic argument for this function
+        let mut rng = StdRng::from_entropy();
+        // Negotiate PRSS first
+        let step = Step::default().narrow(&config.query_type);
+        let prss = negotiate_prss(&gateway, &step, &mut rng).await.unwrap();
+
+        match config.field_type {
+            FieldType::Fp31 => {
+                let ctx = SemiHonestContext::<Fp31>::new(&prss, &gateway);
+                match config.query_type {
+                    #[cfg(any(test, feature = "test-fixture"))]
+                    QueryType::TestMultiply => {
+                        Box::new(test_multiply(ctx, input).await) as Box<dyn Result>
```

Comment:
so can the compiler not figure out how to degrade type to `dyn Result` here?

Reply:
yep it does not auto coerce types into dyn afaik